repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java

```diff
@@ -0,0 +1,100 @@
+package com.epam.esm.repository.impl;
+
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.Order;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.entity.User;
+import com.epam.esm.repository.TagRepository;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.criteria.*;
+import java.util.List;
+import java.util.Objects;
+
+@Repository
+@Transactional
+public class TagRepositoryImpl implements TagRepository {
+    private static final String ORDER_LIST_ATTRIBUTE = "orderList";
+    private static final String CERTIFICATE_LIST_ATTRIBUTE = "giftCertificateList";
+    private static final String TAG_ATTRIBUTE = "tags";
+    private static final String COLUMN_ID = "id";
+    private static final String COLUMN_NAME = "nameTag";
+    private static final int POSITION_WITH_MAX_VALUE = 1;
+    private static final String SQL_FIND_TAG_BY_NAME = "select t.id, t.nameTag from Tag t where t.name = ?";
+
+    @PersistenceContext
+    EntityManager entityManager;
+
+    @Override
+    public List<Tag> findAll(int offset, int limit) {
+        CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Tag> tagCriteriaQuery = criteriaBuilder.createQuery(Tag.class);
+        Root<Tag> root = tagCriteriaQuery.from(Tag.class);
+        tagCriteriaQuery.select(root);
+        return entityManager.createQuery(tagCriteriaQuery)
+                .setFirstResult(offset)
```

Comment:
Seem like OFFSET_DEFAULT_VALUE has been lost here, but I might miss something.

Reply:
Default values for offset and limit have been set in the controller